cpp/include/raft/detail/serialization.hpp

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <raft/core/handle.hpp>
+#include <raft/core/mdarray.hpp>
+
+#include <rmm/device_uvector.hpp>
+
+#include <type_traits>
+#include <vector>
+
+namespace raft::detail {
+
+template <typename T>
+struct serial;
+
+template <typename T, typename... ContextArgs>
+auto call_serialize(uint8_t* out, const T& obj, ContextArgs&&... args) -> size_t
+{
+  return detail::serial<T>::to_bytes(out, obj, std::forward<ContextArgs>(args)...);
+}
+
+template <typename T, typename... ContextArgs>
+auto call_serialize(const T& obj, ContextArgs&&... args) -> std::vector<uint8_t>
+{
+  std::vector<uint8_t> v(
+    call_serialize<T, ContextArgs...>(nullptr, obj, std::forward<ContextArgs>(args)...));
+  call_serialize<T, ContextArgs...>(v.data(), obj, std::forward<ContextArgs>(args)...);
+  return v;
+}
+
+template <typename T, typename... ContextArgs>
+auto call_deserialize(T* p, const uint8_t* in, ContextArgs&&... args) -> size_t
+{
+  return detail::serial<T>::from_bytes(p, in, std::forward<ContextArgs>(args)...);
+}
+
+template <typename T, typename... ContextArgs>
+auto call_deserialize(T* p, const std::vector<uint8_t>& in, ContextArgs&&... args) -> size_t
+{
+  return call_deserialize<T, ContextArgs...>(p, in.data(), std::forward<ContextArgs>(args)...);
+}
+
+template <typename T, typename... ContextArgs>
+auto call_deserialize(const uint8_t* in, ContextArgs&&... args) -> T
+{
+  union res {
```

Comment:
A more idiomatic way to avoid initialization of T would be to have `call_serialize` return `optional<T>` or `expected<T, some_error_code>`.

Reply:
Thanks for the suggestion! Is there an official way to populate the optional using a callback accepting a pointer to the data (i.e. `call_deserialize(T*, ...)`?

I see I haven't explained why I've gone through such a hussle with these many overloads. My goal here is to provide an easy way for other developers to add serialization to new structures. That is, to add serialization to a new structure, one would need to write just two methods:
```
template <T>
struct serial {
  static auto to_bytes(uint8_t* out, const T& obj) -> size_t { ... }
  static auto from_bytes(T* p, const uint8_t* in) -> size_t { ... }
};
```
Although I love the return-value approach, I'm afraid I cannot use it here, because some of the classes in raft and cuml do not have some of the copy, move, or nullary constructors. Hence I've come up with this idea that it should be the responsibility of the serialization implementer to initialize the object in the `from_bytes` method.

The rest, all `call_serialize`/`call_deserialize` are implemented in terms of these two. At this moment, we don't seem to have a consensus over which target type to use for serialization - so I added a few of them here.

Assuming that `from_bytes` is available as-is, the union approach is the only way I could come up to return the value in this overload. I'd be happy to use `optional` for this, if it's possible!

Reply:
@achirkin Thanks for explaining!  I see the difficulty now in using `optional`.  Another way to express this besides a `union` would be to allocate an array of bytes with the same size and alignment as `T`, and then placement new into it.  Here is an example.
```c++
class NotDefaultConstructible {
public:
  NotDefaultConstructible() = delete;
  NotDefaultConstructible(int i) : i_(i) {}

  int value() const { return i_; }

private:
  int i_;
};

void deserialize_raw(NotDefaultConstructible* out, std::string_view s)
{
    int value{};  
    auto result = std::from_chars(s.data(), s.data()+s.size(), value);
    if (result.ec == std::errc::invalid_argument) {
        throw std::invalid_argument{"String does not represent an integer"};
    }
    else if (result.ec == std::errc::result_out_of_range) {
        throw std::out_of_range{"String represents an integer, but it does not fit in int"};
    }
    // WARNING: assumes out does NOT point to a valid object.
    new(out) NotDefaultConstructible(value);
}

int main()
{
    constexpr size_t num_bytes = sizeof(NotDefaultConstructible);
    alignas(num_bytes) std::byte ptr[num_bytes];

    NotDefaultConstructible* out = reinterpret_cast<NotDefaultConstructible*>(ptr);
    deserialize_raw(out, s);

    std::cout << "Value: " << out->value();
    out->~NotDefaultConstructible(); // oof, we have to do this by hand

    return 0;
}
```

Reply:
Ouch, I don't want to call the destructor by hand in the presence of exceptions in `deserialize_raw` :) I guess I'd better go with the `union` approach if you don't see anything criminal in that.